### PR TITLE
bit testing and setting functions

### PR DESCRIPTION
**Pull request by burner**
Bit Testing Functions
---------------------
```d
bool testBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);

bool testAnyBit(T)(const T bitfield) if(isIntegral!T);

bool testNoBit(T)(const T bitfield) if(isIntegral!T);

bool testAllBit(T)(const T bitfield) if(isIntegral!T);
```

Bit Setting Functions
---------------------
```d
T setBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);

T setBit(T)(const T bitfield, const ulong idx, bool value) if(isIntegral!T);

T flipBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);

T resetBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);
```

Bit Pretty Printing
----------------------------
```d
void bitsPrettyPrint(T,Out)(const T value, ref Out output)
	if (isIntegral!T && isOutputRange!(Out,string));

string bitsPrettyPrint(T)(const T value) if (isIntegral!T);

```

github: https://github.com/burner/BitFiddle
dub: http://code.dlang.org/packages/bitfiddle

### CODE REVIEW

**Comment by burner** [comment]
These are very short functions but getting them right thanks to value range propagation is a pain. Now you just have to call a function.

**Comment by burner** [comment]
@klickverbot it inserts the right casts at the right spots so you don't have to remember that byte become ints when you bitwise (or|and) them. of cause some people know, but calling one of those functions is way easier.

**Comment by 9il** [comment]
Maybe Bit Setting Functions is enough?

File: std/bitmanip.d [hunk 1]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+}
+
+///
+unittest
+{
+    int a = 0b0000_0000_0010_0001;
+    assert(testBit(a, 0));
+    assert(!testBit(a, 1));
+    assert(testBit(a, 5));
+    assert(!testBit(a, 6));
```
Comment (wilzbach) [inline std/bitmanip.d]: Needs tests for more types.

File: std/bitmanip.d [hunk 2]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+If `idx >= T.sizeof * 8` the bahaviour is undefined.
+
+Params:
+   bitfield = the integer containing bit to test
+   idx = the index of the bit to test
+
+Returns: `true` if the bit is set, `false` otherwise.
+*/
+bool testBit(T)(const T bitfield, const ulong idx) if (isIntegral!T)
+{
```
Comment (wilzbach) [inline std/bitmanip.d]: Needs input contract to avoid testing for invalid bits

File: std/bitmanip.d [hunk 3]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+returned copy.  If `idx >= T.sizeof * 8` the bahaviour is undefined.
+
+Params:
+   bitfield = the integer to modify
+   idx = the index of the bit to set
+
+Returns: a copy of `bitfield` with bit `idx` set
+*/
+T setBit(T)(const T bitfield, const ulong idx) if (isIntegral!T)
+{
```
Comment (wilzbach) [inline std/bitmanip.d]: A templated version here if the bit is known at CTFE would be  faster?

File: std/bitmanip.d [hunk 4]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+    }
+}
+
+}
+
+/** Converts a integer bitfield into a four bit at a time representation,
+where every four bits are seperated by an underscore.
+*/
+void bitsPrettyPrint(T,Out)(const T value, ref Out output)
+    if (isIntegral!T && isOutputRange!(Out,string))
```
Comment (wilzbach) [inline std/bitmanip.d]: Would it make sense to move this pretty bit printing to std.format? E.g. `%B`

File: std/bitmanip.d (hunk 3 above)
Comment (burner) [inline std/bitmanip.d]: that would be overkill IMO

File: std/bitmanip.d (hunk 4 above)
Comment (burner) [inline std/bitmanip.d]: yes, I will work on that.

File: std/bitmanip.d (hunk 2 above)
Comment (burner) [inline std/bitmanip.d]: really, I mean if you don't knoe how many bits there are in an int even those functions won't help you ;-)

**Comment by 9il** [comment]
This PR looks like overengineering, imho

**Comment by burner** [comment]
OK, so what about just keeping

```d
bool testBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);
T setBit(T)(const T bitfield, const ulong idx, bool value) if(isIntegral!T);
T flipBit(T)(const T bitfield, const ulong idx) if(isIntegral!T);
```

and adding an %B to format so it pretty prints like the bits as my prettyPrint function does?

**Comment by 9il** [comment]
Yes, this looks good for me

**Comment by 9il** [comment]
Why `idx` is not uint?

**Comment by burner** [comment]
one less cast

**Comment by 9il** [comment]
Why `ulong` is default integer type?

**Comment by burner** [comment]
I do not understand the question?

**Comment by 9il** [comment]
I mean that this will be bad code for 32 bit DMD:
 +    ulong value = cast(ulong) bitfield;
 +    ulong mask = 1UL << idx;
 +    value |= mask;
 +    return cast(T) value;

**Comment by burner** [comment]
If that is bad code  (I assume by bad code you, will have a bug, not bad codegen), than the autotester is broken. Because I test that case and the autotester does not complain.

**Comment by 9il** [comment]
Sorry for my explanation. I mean bad code optimisation

**Comment by burner** [comment]
what do you mean by bad code optimization?

**Comment by 9il** [comment]
> what do you mean by bad code optimization?
>  Hide all checks

32 bit CPU has not 64 bit integer registers. So using `ulong` for all types is not optimal

**Comment by burner** [comment]
I get it now. I'll work on making the cast to uint if possible

File: std/bitmanip.d (hunk 2 above)
Comment (wilzbach) [inline std/bitmanip.d]: well your bitfield could be a `long` and then you accidentally change it to `int` - I think it's always a good style to check for invalid input ;-)

File: std/bitmanip.d (hunk 3 above)
Comment (wilzbach) [inline std/bitmanip.d]: Why? It's faster and usually if one uses bitfields one _really_ does care about performance.
I measured this quickly with my [d-benchmarks](https://github.com/wilzbach/d-benchmarks) (e.g. `make setbit.all`). For both `ldc` and `dmd` you get about a 10% increase when using CTFE (of course with all optimizations) - I think that's worth it ;-)

Comment (burner) [inline std/bitmanip.d]: Of course it is faster, I never doubted that. But how often do you think people will have a function where they know at CT what bit they want to set, but the value they want to set it in is a RT value. And even then people can use this and CTFE to build a bitmask and than just | or & 

Additionally, someone already said this PR was overengineering, what chance of getting this in do you think will be there if I add partial CTFE versions of the functions \(\lim_{x = 0}\)

Comment (klickverbot) [inline std/bitmanip.d]: The difference should disappear with inlining.

Comment (burner) [inline std/bitmanip.d]: lets see what the big boss has to say @andralex

Comment (wilzbach) [inline std/bitmanip.d]: > The difference should disappear with inlining anyway. (Still not a fan of having them in the first place. ^^)

@klickverbot I tried both [separate files] (https://github.com/wilzbach/d-benchmarks) and [one single file](https://gist.github.com/wilzbach/c8dd833fd6f2df43604b9c6b0cd0dff4) and the CTFE difference doesn't disappear.

> Of course it is faster, I never doubted that. But how often do you think people will have a function where they know at CT what bit they want to set, but the value they want to set it in is a RT value.  And even then people can use this and CTFE to build a bitmask and than just | or &

I thought that's the common case as the programmer usually knows his flags at CT, but wants to test/set them at RT.

> Additionally, someone already said this PR was overengineering, what chance of getting this in do you think will be there if I add partial CTFE versions of the functions $\lim_{x = 0} $

Understood :/

**Comment by wilzbach** [comment]
@burner btw what about [core.bitop](https://dlang.org/phobos/core_bitop.html)?

- bt - tests the bit
- btc - test & complement
- btr - test & resets (0)
- bts - test & sets (1)

**Comment by burner** [comment]
they work only on size_t*

File: std/bitmanip.d [hunk 5]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
 }
+
+pure @safe @nogc nothrow {
+
+private template upTo(T)
+{
+    enum upTo = T.sizeof * 8UL;
+}
+
+private template minUnsignedIntegral(T)
```
Comment (andralex) [inline std/bitmanip.d]: What's wrong with just using `Unsigned!T`?

File: std/bitmanip.d [hunk 6]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+/** Tests if any bit is set in the integer `bitfield`.
+
+Params:
+   bitfield = the integer to test
+
+Returns: `true` if any bit is set, `false otherwise.
+*/
+bool testAnyBit(T)(const T bitfield) if (isIntegral!T)
+{
+    return bitfield != T.init;
```
Comment (andralex) [inline std/bitmanip.d]: Really? I have to throw the towel here. A function that tests whether any bit is set within an integral is an insult to the potential user's intelligence.

File: std/bitmanip.d [hunk 7]
```diff
@@ -4005,3 +4005,444 @@ auto bitsSet(T)(T value) @nogc pure nothrow
...
+/** Tests if no bit is set in the integer `bitfield`.
+
+Params:
+   bitfield = the integer to test
+
+Returns: `true` if no bits are set, `false otherwise.
+*/
+bool testNoBit(T)(const T bitfield) if (isIntegral!T)
+{
+    return bitfield == T.init;
```
Comment (andralex) [inline std/bitmanip.d]: somebody shoot me

**Comment by andralex** [comment]
This kind of stuff should not be in Phobos.

**Comment by JackStouffer** [comment]
@andralex For a module called `bitmanip`, a bit setting function seems to be a natural fit, no? The `setBit` and `flipBit` functions seem to be useful utilities to save the user the time they would spend using trial and error to get the correct result from int promotion.

I think that closing this was a bit premature.

**Comment by andralex** [comment]
I'll reopen for further discussion, but one-liners such as `setBit` and `flipBit` don't hold much promise.

**Comment by WalterBright** [comment]
I'm sorry this has been a fair amount of work for you guys, in the light of what I'm about to say. But it falls into something I've argued against for years - a mass of trivial one-liners that take more time to read the documentation than just using & | and ^ operators. Avoiding learning about the trivial integral promotion rules is not a justification for this - and this is much, much more to learn than that - and the programmer will still have to learn about integral promotion. And if & | ^ are too hard for a user and must be put in a library, are + - * / next?

441 lines of source text to support 8 one liners and 2 trivial functions is an indication as well that something has gone awry.

This needs a much more compelling rationale, but I agree with Andrei that it's hard to see much promise in them. I apologize for not noticing this sooner, I wish I could have saved you the effort.

**Comment by wilzbach** [comment]
> This needs a much more compelling rationale, but I agree with Andrei that it's hard to see much promise in them. I apologize for not noticing this sooner, I wish I could have saved you the effort.

FWIW I will try to give a rationale:

1) While I agree that one-liners take more time to find than to write, is this really nice code to read?

```d
cast(T)(cast(ulong)(bitfield) ^ (-cast(ulong)(value) ^ cast(ulong)(bitfield)) & (1UL << 8));
```

... instead of:

```
setBit!8(bitfield, 1)
```

2) There's a [performance difference](https://github.com/dlang/phobos/pull/4703#discussion_r73625483) if the mask can be created at CTFE
3) Bit pretty-printing is something very helpful (but that will be a separate PR for `std.format`)
4) It's already part of our library API - [core.bitop](https://dlang.org/phobos/core_bitop.html) provides convenience wrapper, but they have such ugly names than no one can find them (& they just work on `size_t*`)

**Comment by 9il** [comment]
> cast(T)(cast(ulong)(bitfield) ^ (-cast(ulong)(value) ^ cast(ulong)(bitfield)) & (1UL << 8));

This example has a lot of pointless casts.

I agree with @WalterBright

**Comment by andralex** [comment]
The corollary here is that the set of people who simultaneously need to set bits in integrals and don't know how to is near empty.

File: std/bitmanip.d (hunk 3 above)
Comment (klickverbot) [inline std/bitmanip.d]: @wilzbach:

> (my benchmarks may be flawed though)

Your tests try hard to make the code non-inlinable. Of course there will be a performance difference then.

Comment (klickverbot) [inline std/bitmanip.d]: > one single file 

Replace `idx` in the runtime version with the literal `10` (or make `idx` an enum). The performance difference disappears, then.

Comment (wilzbach) [inline std/bitmanip.d]: > Your test harness tries hard to make the code non-inlinable. Of course there will be a performance difference then.
> Replace idx in the runtime version with the literal 10 (or make idx an enum). The performance difference will disappear, then.

I thought that's what people call "bias"? :P 
Thanks!

Comment (klickverbot) [inline std/bitmanip.d]: Bias on your side or mine? ;P

I suppose my point is that while it is of course possible to accidentally write slower code with the runtime interface, it's not that hard to ensure that the value is actually a compile-time constant at the call site if you care about it. (You can also always write your private template wrapper to ensure that.)

Comment (wilzbach) [inline std/bitmanip.d]: > Bias on your side or mine? ;P

I thought that's obvious ;-)

> I suppose my point is that while it is of course possible to accidentally write slower code with the runtime interface, it's not that hard to ensure that the value is actually a compile-time constant at the call site if you care about it.

I just didn't know that the compiler is already so smart!

Comment (klickverbot) [inline std/bitmanip.d]: > I just didn't know that the compiler is already so smart!

For future reference, inlining of tiny functions and scalar constant propagation is actually the one thing you can pretty much be sure will happen.

**Comment by andralex** [comment]
>I should have probably phrased my comment less politely.

Instead, I should have phrased my comment more politely. Thanks for being civil, my way of putting things is not called for.

**Comment by burner** [comment]
> The corollary here is that the set of people who simultaneously need to set bits in 
> integrals >and don't know how to is near empty.
[citation needed]

@WalterBright I disagree, algebra and integer promotion makes sense, bit fiddling and integer promotion does not. That is why I created these functions. I properly got carried away with some functions, but setting and testing individual bits via functions makes room in my harddrive to focus on more elaborate tasks.

**Comment by WalterBright** [comment]
```
cast(T)(cast(ulong)(bitfield) ^ (-cast(ulong)(value) ^ cast(ulong)(bitfield)) & (1UL << 8));
```
Only one cast to ulong is needed, the rest will automatically happen. Library functions are not a solution to not understanding integral promotions rules - they are a fundamental, basic feature of the language, not some weird corner case.

core.bitop exists for two reasons:

1. as mentioned before, ensure that the compiler generates the special instructions for them

2. the bit test functions work on arrays that are more than one machine word (as do their corresponding special instructions)

**Comment by burner** [comment]
> Library functions are not a solution to not understanding integral promotions rules

Actually, they are!

For normal math you don't need to know integer promotion rules, they just work 99.999% of the time. And that is why they are good. For bit fiddling you need to know, or you use these functions.

**Comment by andralex** [comment]
Sadly I don't see anything salvageable in here. It would be nice to migrate the bits printer to `format` as a flag for formatting integrals. I suggest closing this.
